### ISSUE TITLE
Bot Feature Requirements

### ISSUE THREAD

**Issue by lealgamb** [vocatwk/CodeU-Nemo#70]
When a user @ mentions the bot (i.e. @NemoBot) in a message, the bot should respond.

Bot Features:

- Provide keywords a user can mention 

- Provide answers to user questions, kind of like a Q&A feature.

- Provide statistics for conversations (i.e. number of users and messages in conversations)

- Deal with invalid user input (i.e. "Sorry, I didn't understand that. Try mentioning one of these words [...]"

These are the initial features I can think of. Let me know what you guys think.

**Comment by yebrahim** [comment]
Good start. Can we try to add more details?
- What kind of questions can the user ask? A specific initial list would be good here, so that the implementation has something to start with.
- I'm assuming we'll start with a hard-coded list of questions, where the user's query must exactly match one of the questions to be considered, can we discuss the list here?

The goal of making this specific is to guide the implementation process, by making the end goal as clear as possible, so that we can then decide on the step breakdown (prototype, MVP, or any other steps we want to add).

**Comment by lealgamb** [comment]
I've never used natural language processing but I'm definitely willing to try to help and learn along the way. I like the reminder feature and it can pair up nicely with the notifications. 

Also I edited the original post with some keywords. @all

**Comment by yebrahim** [comment]
We can look into using the DialogFlow API as a stretch goal here. let's see when we get there. :)

**Comment by vocatwk** [comment]
Is it so that users can add bots that other users can use? and so users can add questions and answers to an already exixsting bot to enhance it?

**Comment by yebrahim** [comment]
Well, since this is an open source project, I imagine anyone interested might want to contribute to the code to add or modify a bot. If we want to support developer contributions to bots (or our own in the future), we should try to make the framework extensible so that when adding a new bot, the developer does not need to understand too many pieces in the codebase.

**Comment by lealgamb** [comment]
That's a good idea. I was thinking of storing the questions/keywords in a map. On creation, each Bot has the "essential" information in a map. Public functions like `addBotEntry(Key k, Value v)` can be implemented with perhaps a Bot Management page?

**Comment by yebrahim** [comment]
Yes, I think this is reasonable. The workflow for adding a new bot is then to create the class (perhaps extend some base `Bot` class?) and implement some method such as `parseMessage` or `answerMention`, then maybe add it to a singleton map with the appropriate mention name?

**Comment by lealgamb** [comment]
Sorry I haven't worked with singleton maps before. Before I ask about it, I just want to make sure I understand what you described.

- `UserCreatedBot` extends `Bot`

- `parseMessage` or `answerMention` will perform the same functionality that the current `Bot` class has

- Add it to a singleton map. Will the key be the mention name and the value be `UserCreatedBot`?

Also, from what I understand we can't put items in a singleton map later on, so I'm having trouble figuring out why we want this.

**Comment by lealgamb** [comment]
I was referring to [this SO question](https://stackoverflow.com/questions/7125536/when-would-i-use-java-collections-singletonmap-method), but I guess it's different from what we're talking about.

This cleared things up, thanks. So just to be sure, we'll have one "main" bot that's responsible for answering messages about our app. Then, users will have the ability to create their own bots. At this stage, will the users be responsible for creating the message->answer map, or is this something we should handle, perhaps with the DialogFlow API?

**Comment by yebrahim** [comment]
I think when we have such a framework in place (the singleton map controller), we should rework the current `Bot` class implementation to name it `NemoBot` for example or `HelpBot`, and have it extend from the new `Bot` base class (or just interface if there's no common functionality).
Thoughts?

**Comment by lealgamb** [comment]
Renaming it makes sense. And I think making it extend the base class will be better since all bots will have to parse a message and answer a mention.

**Comment by lealgamb** [comment]
Hmmm, I can't think of any. And you're right, I was only thinking of it at an abstract level. So the interface does make more sense because all bots should behave in a similar way, but how they do so can vary. Am I thinking of this correctly?

**Comment by yebrahim** [comment]
Yes.
The difference between using an interface and a base class is that the latter can contain logic. Let's say we want a `greet` method which, when called, returns a message: "Hello, this is HelpBot!". The code to achieve this does not differ from one bot implementation to the next (it can get the bot name from an attribute), it's just a function that returns a static message. In this case, it would make sense to have a base class that has a method `greet`, which our framework can call.
If we want bot classes to be able to implement all methods differently, however, then what we need is an interface. It's a looser way to make sure that all instances adhere to the same API.
